spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/actions/BaseSparkAction.java

```diff
@@ -176,16 +186,35 @@ protected Dataset<Row> loadMetadataTable(Table table, MetadataTableType type) {
     return SparkTableUtil.loadMetadataTable(spark, table, type);
   }
 
-  private static class ReadManifest implements FlatMapFunction<ManifestFileBean, String> {
-    private final Broadcast<FileIO> io;
+  private static class ReadManifest implements FlatMapFunction<ManifestFileBean, Tuple2<String, String>> {
+    private final Broadcast<Table> table;
 
-    ReadManifest(Broadcast<FileIO> io) {
-      this.io = io;
+    ReadManifest(Broadcast<Table> table) {
+      this.table = table;
     }
 
     @Override
-    public Iterator<String> call(ManifestFileBean manifest) {
-      return new ClosingIterator<>(ManifestFiles.readPaths(manifest, io.getValue()).iterator());
+    public Iterator<Tuple2<String, String>> call(ManifestFileBean manifest) {
+      return new ClosingIterator(entries(manifest));
+    }
+
+    public CloseableIterator<Tuple2<String, String>> entries(ManifestFileBean manifest) {
+      switch (manifest.content()) {
+        case DATA:
```

Comment:
isn't this always DATA  in current imiplementation?

Reply:
What do you mean?  There are now manifest with two kind of ManifestContent, and both will get read by expireSnapshot scan to find the files to be deleted.

Reply:
ManifestFileBean.content is always returning DATA looking from src

Reply:
Good catch, this seems it is working by chance, ManifestFiles.read() only checks to make sure content=Data, but doesn't use the content anywhere else in this path.  The file returned can still be delete or data file.

Not sure if it's worth to implement manifests.content field (I guess that's what was done in the original pr: #2518).   if there is not much use now, maybe we can do it in a separately and here just add a new method to ManifestFiles.files() that returns a Pair of type/path, similar to ManifestFiles.paths().  @aokolnychyi any preference?

Reply:
Yea after thinking, another option is to get replace  this altogether with a query on 'all_files' table, which should be coming soon.  It's a cleaner way as the complexity is kept within the metadata table, though requires to wait until that one.

Reply:
I am afraid we don't project `content` in manifests-related metadata tables but we should probably fix that. I discovered that in another use case. I can submit a PR for that.

Reply:
For reference, related pr : https://github.com/apache/iceberg/pull/2510

Reply:
Ref: performance, yea I realized that all_entries table is not parallel in reading manifest-lists while commenting on #3457, sorry forgot to update here.